packages/alfa-rules/src/sia-er62/rule.ts

```diff
@@ -69,54 +61,112 @@ export default Rule.Atomic.of<Page, Element>({
   requirements: [Criterion.of("1.4.1")],
   tags: [Scope.Component, Stability.Experimental, Version.of(2)],
   evaluate({ device, document }) {
+    // Contains links (key) and their containing paragraph (value)
     let containers: Map<Element, Element> = Map.empty();
 
     return {
       applicability() {
-        return visit(document, None);
+        // Contains links (key) and the parents of the textnodes the links contain (value)
+        let linkText: Map<Element, Set<Element>> = Map.empty();
+        // Contains containers (key) without link descendants and the parents of the textnodes the containers have (value)
+        let nonLinkText: Map<Element, Set<Element>> = Map.empty();
+
+        gather(document, None, None);
+        return getApplicableLinks();
 
-        function* visit(
+        function gather(
           node: Node,
-          container: Option<Element>
-        ): Iterable<Element> {
+          container: Option<Element>,
+          link: Option<Element>
+        ): void {
           if (isElement(node)) {
-            // If the element is a semantic link, it might be applicable.
-            if (
-              test(
-                hasRole(device, (role) => role.is("link")),
-                node
-              )
-            ) {
-              if (
-                container.isSome() &&
-                node
-                  .descendants({ flattened: true })
-                  .some(and(isText, isVisible(device)))
-              ) {
-                containers = containers.set(node, container.get());
-                return yield node;
-              }
+            const isLink = hasRole(device, (role) => role.is("link"));
+            const isParagraph = and(
+              hasRole(device, "paragraph"),
+              hasNonLinkText(device)
+            );
+
+            if (container.isSome() && isLink(node)) {
+              // For each link, know store its containing paragraph
+              containers = containers.set(node, container.get());
+              link = Option.of(node);
             }
 
-            // Otherwise, if the element is a <p> element with non-link text
+            // Otherwise, if the element is a paragraph element with non-link text
             // content then start collecting applicable elements.
-            else if (
-              test(
-                and(hasRole(device, "paragraph"), hasNonLinkText(device)),
-                node
-              )
-            ) {
+            if (isParagraph(node)) {
               container = Option.of(node);
             }
           }
 
+          const isTextNode = test(and(isText, isVisible(device)), node);
+          const parent = node.parent();
```

Comment:
```suggestion
          const parent = node.parent().filter(isElement);
```
It will be safer to keep that as part of the guard and avoid the risk of getting from `none` later.